Perf guidelines (#303)
MicrosoftDocs/OData-docs: Odata-docs/performance/disable-writer-validations.md

---
title:  "Disable writer validations"
description: A guide on how to improve performance by disabling writer validations.
date:   2024-01-23
ms.date: 1/23/2023
author: habbes
ms.author: clhabins
---

# Disable writer validations

**Applies To**:[!INCLUDE[appliesto-webapi](../includes/appliesto-webapi-v8.md)][!INCLUDE[appliesto-webapi](../includes/appliesto-webapi-v7.md)][!INCLUDE[appliesto-odatalib](../includes/appliesto-odatalib-v7.md)]

During serialization, the OData writer performs various validations to ensure that the payload is valid (e.g. no duplicate properties) and conforms to the schema (e.g. property type matches the entity type's definition). These validations introduce some overhead to the serialization process. If you know ahead of time that the payload you're writing is correct, you can disable one or more validations to improve performance.

Use the [`ODataMessageWriterSettings.Validations`](/dotnet/api/microsoft.odata.odatamessagewritersettings.validations) property to configure which validations you want to enable or disable.

> You can disable all validations using [`ValidationKinds.None`]:

Reviewer: ```suggestion
You can disable all validations using `ValidationKinds.None` flag:
```
The square brackets look odd in doc preview

Author: Thanks. This was supposed to be a link

Author: I've opted to link to the `ValidationKinds` enum page instead.

```csharp
messageWriterSettings.Validations = ValidationKinds.None;
```

You can selectively enable validations using the `&` operator:

```csharp
messageWriterSettings.Validations = ValidationKinds.ThrowOnDuplicatePropertyNames & ValidationKinds.ThrowIfTypeConflictsWithMetadata
```

> You can enable all validations using [`ValidationKinds.All`]:

Reviewer: ```suggestion
You can enable all validations using `ValidationKinds.All` flag:
```
The square brackets look odd in doc preview

Author: Actually I was suppost to link them to the relevant documentation page. Thanks for catching that.

Author: Removed the square brackets since I couldn't find the actual reference link.

Author: I've opted to link to the `ValidationKinds` enum page instead.

```csharp
messageWriterSettings.Validations = ValidationKinds.All;
```

> To learn more about different validation options, consult the [`ValidationKinds` enum documentation](/dotnet/api/microsoft.odata.validationkinds).

Reviewer: ```suggestion
To learn more about different validation options, check out the [`ValidationKinds` enum documentation](/dotnet/api/microsoft.odata.validationkinds).
```

> ## Disabling writer validations when using `Microsoft.OData.Core` directly

Reviewer: ```suggestion
## Disabling writer validations when using `Microsoft.OData.Core` library directly
```

> When using the core library directly, you pass the `ODataMessageWriterSettings` directly to the [`ODataMessageWriter`](/dotnet/api/microsoft.odata.odatamessagewriter) construtor.

Reviewer: How come `ODataMessageWriterSettings` gets to be code-highlighted while `ODataMessageWriter` doesn't? How come `ODataMessageWriter` gets to be linked to the library documentation but `ODataMessageWriterSettings` doesn't?

Author: They're now both code-formatted and linked to the respective reference pages.

```csharp
ODataMessageWriterSettings settings = new ODataMessageWriterSettings
{
    ODataUri = new ODataUri() { /* ... */ },
    Validations = ValidationKinds.None
};

ODataMessageWriter writer = new ODataMessageWriter(odataMessage, settings, edmModel);

// ...
```

## Disabling writer validations when using `Microsoft.AspNetCore.OData` 8

In `Microsoft.AspNetCore.OData` 8, you can pass a service configuration action as the third argument of `AddRouteComponents`. Use

> this to inject a custom request-scoped `ODataMessageWriterSettings`, with validations disabled, to the service container:

Reviewer: ```suggestion
this to inject a custom request-scoped `ODataMessageWriterSettings` - with the desired validation options - to the service container:
```

```csharp
services
.AddControllers()
.AddOData(options =>

>     options.AddRouteComponents("routePrefix", edmModel, container =>

Reviewer: Using named parameters would make this snippet self-describing and easier to understand, i.e.,
```csharp
routePrefix: "...",
model: edmModel,
configureAction: container => // ...
```

    {
        container.AddScoped<ODataMessageWriterSettings>(_ => new ODataMessageWriterSettings
        {
            Validations = ValidationKinds.None
        })
    }));
```

> ## Disabling writer validations when using `Microsoft.AspNetCore.OData` 7

Reviewer: ```suggestion
## Disabling writer validations when using `Microsoft.AspNetCore.OData` 7 library
```

In `Microsoft.AspNetCore.OData` 7, you can pass a service configuration action to the `MapODataRoute` and `MapODataServiceRoute` methods.

> Use this to inject a custom request-scoped ODataMessageWriterSettings` to the service container.

Reviewer: ```suggestion
Use this to inject a custom request-scoped `ODataMessageWriterSettings` to the service container.
```

### Using endpoint routing

```csharp
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
{

    // ...

    app.UseEndpoints(endpoints =>
    {
        endpoints.MapODataRoute(

>             "odata", “odata”,

Reviewer: Named parameters might improve this snippet and make it easier to understand, i.e.,
```
routeName: "odata",
routePrefix": "odata",
configureAction: container => // ...
```

            container =>
            {

>                 container.AddService<ODataMessageWriterSettings>(Microsoft.OData.ServiceLifetime.Scope, _ => new ODataMessageWriterSettings

Reviewer: ```suggestion
                container.AddService<ODataMessageWriterSettings>(Microsoft.OData.ServiceLifetime.Scoped, _ => new ODataMessageWriterSettings
```

                {
                    Validations = ValidationKinds.None
                });

>                 container.AddService(Microsoft.OData.ServiceLifetime.Singleton, sp => model);

Reviewer: ```suggestion
                container.AddService(Microsoft.OData.ServiceLifetime.Singleton, _ => model);
```
Because you used `_` 4 lines above....

                container.AddService<IEnumerable<IODataRoutingConvention>>(Microsoft.OData.ServiceLifetime.Singleton,

>                     sp => ODataRoutingConventions.CreateDefaultWithAttributeRouting("nullPrefix", endpoints.ServiceProvider));

Reviewer: Where did "endpoints" come from? Did you mean `sp.ServiceProvider` or something else?

>                     sp => ODataRoutingConventions.CreateDefaultWithAttributeRouting("nullPrefix", endpoints.ServiceProvider));

Reviewer: Can you pass `sp` in place of `endpoints.ServiceProvider`?

>                     sp => ODataRoutingConventions.CreateDefaultWithAttributeRouting("nullPrefix", endpoints.ServiceProvider));

Reviewer: The route prefix you specified several lines above is "odata", Maybe use the same one here to avoid possibly misleading the reader...

            });
    });

    // ...
}
```

> ### Using Endpoint Routing

Reviewer: The previous level 3 heading was also "Using endpoint routing"...
I'm not sure there's that much additional value in including a code sample for both Mvc and Endpoint routing. The code samples don't differ by much other than for the method being called. I think I'd get rid of one and then there would be no need for the level 3 headings.

Author: I've renamed this to "Using MVC routing". I figured it wouldn't hurt to show the code sample here for completeness. Since it's already here anyway.

Use the [`ODataMessageWriterSettings.Validations`](/dotnet/api/microsoft.odata.odatamessagewritersettings.validations) property to configure which validations you want to enable or disable.

You can disable all validations using [`ValidationKinds.None`](/dotnet/api/microsoft.odata.validationkinds):

```csharp
messageWriterSettings.Validations = ValidationKinds.None;
```

You can selectively enable validations using the `&` operator:

```csharp
messageWriterSettings.Validations = ValidationKinds.ThrowOnDuplicatePropertyNames & ValidationKinds.ThrowIfTypeConflictsWithMetadata
```

You can enable all validations using [`ValidationKinds.All`](/dotnet/api/microsoft.odata.validationkinds):

```csharp
messageWriterSettings.Validations = ValidationKinds.All;
```

To learn more about different validation options, consult the [`ValidationKinds` enum documentation](/dotnet/api/microsoft.odata.validationkinds).

## Disabling writer validations when using `Microsoft.OData.Core` library directly

> When using the core library directly, you pass the [`ODataMessageWriterSettings`](/dotnet/api/microsoft.odata.odatamessagewritersettings) to the [`ODataMessageWriter`](/dotnet/api/microsoft.odata.odatamessagewriter) construtor.

Reviewer: ```suggestion
When using the core library directly, you pass the [`ODataMessageWriterSettings`](/dotnet/api/microsoft.odata.odatamessagewritersettings) to the [`ODataMessageWriter`](/dotnet/api/microsoft.odata.odatamessagewriter) constructor.
```

```csharp
ODataMessageWriterSettings settings = new ODataMessageWriterSettings
{
    ODataUri = new ODataUri() { /* ... */ },
    Validations = ValidationKinds.None
};

ODataMessageWriter writer = new ODataMessageWriter(odataMessage, settings, edmModel);

// ...
```

## Disabling writer validations when using `Microsoft.AspNetCore.OData` 8 library

In `Microsoft.AspNetCore.OData` 8, you can pass a service configuration action as the third argument of `AddRouteComponents`. Use
this to inject a request-scoped instance of `ODataMessageWriterSettings` - with the desired options - to the service container:

```csharp
services
.AddControllers()
.AddOData(options =>
    options.AddRouteComponents(
        routePrefix: "routePrefix",
        model: edmModel,
        configurateServices: container =>
        {
            container.AddScoped<ODataMessageWriterSettings>(_ => new ODataMessageWriterSettings
            {
                Validations = ValidationKinds.None
            })
        }));
```

> ## Disabling writer validations when using `Microsoft.AspNetCore.OData` 7 library

Reviewer 2: Is this `Microsoft.AspNetCore 7 only or even Microsoft.AspNet.OData`

Author: Yes. `ODataUtf8JsonWriter` is not supported in `Microsoft.AspNet.OData` because it depends on `Utf8JsonWriter` which is not supported in .NET Framework.

Reviewer: @habbes I don't think this particular article is related to `ODataUtf8JsonWriter`, is it?

Author: Hmm. You are correct. I had confused this with a different conversation. Let me revert back with a response.

Author: You can do this in AspNet.OData 7 as well using the `MapODataServiceRoute` method like the MVC routing approach below. The only difference is that the AspNet 7.x configuration setup looks different from the AspNetCore one (there's no Startup class with a Configure method, there's a `WebApiConfig` class with a `Register` method that takes a `HttpConfiguration` argument.

But the OData part is identical. I think readers would still be able to easily adapt this sample to their AspNet.OData code.

In `Microsoft.AspNetCore.OData` 7, you can pass a service configuration action to the `MapODataRoute` and `MapODataServiceRoute` methods.
Use this to inject a request-scoped instance of `ODataMessageWriterSettings` to the service container.

### Using endpoint routing

```csharp
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
{

    // ...

    app.UseEndpoints(endpoints =>
    {
        endpoints.MapODataRoute(
            routeName: "odata",
            routePrefix: “odata”,
            configureAction: container =>
            {
                container.AddService<ODataMessageWriterSettings>(Microsoft.OData.ServiceLifetime.Scoped, _ => new ODataMessageWriterSettings
                {
                    Validations = ValidationKinds.None
                });

                container.AddService(Microsoft.OData.ServiceLifetime.Singleton, _ => model);
                container.AddService<IEnumerable<IODataRoutingConvention>>(Microsoft.OData.ServiceLifetime.Singleton,
                    sp => ODataRoutingConventions.CreateDefaultWithAttributeRouting("odata", sp));
            });
    });

    // ...
}
```

> ### Using MVC Routing

Reviewer: nit:
```suggestion
### Using MVC routing
```
Considering casing used for other headings
